### ISSUE TITLE
Evaluating the neural network in the loss function multiple times is not being accounted for

### ISSUE THREAD

**Issue by timudk** [pytorch/opacus#461]
## 🐛 Bug

When evaluating the neural network N times in the loss function, Opacus does not take account of this internally. The internal DP-SGD seems to sum all N gradients but does not divide by N.

## Please reproduce using our [template Colab](https://colab.research.google.com/drive/1R-dnKipK9LOVV4_oKbuHoq4VvGKGbDnd?usp=sharing) and post here the link
https://colab.research.google.com/drive/1DWcZXHDPz2ROC5XLkw21mSyhpVfM0zHQ?usp=sharing

## To Reproduce
> :warning: **We cannot help you without you sharing reproducible code**. Do not ignore this part :)

1. Run the code once with N=1 (standard), obtaining 92.4 % test acc
2. Run the code again with N>1 (adversarial behavior), I set N=100 and obtain 100% test acc

N can be modified in the line above "privacy_engine.make_private"

## Expected behavior

Internally, it should be taken care of that the NN is evaluated N times on the same data points. Therefore, summed gradients and clipped gradients should be divided by N*B (B being the original batch size) rather than B.

**Comment by alexandresablayrolles** [comment]
Thanks for raising this issue @timudk. Opacus assumes each input is a new data point, so it indeed will be "fooled" if the same input is repeated N times. While there is an easy fix to this exact problem (GradSampleModule can check for equality of multiple inputs), it does not really solve the general problem (inputs can be the same up to data augmentation).
At the end of the day, it is not possible for Opacus to prevent adversarial behavior. We rather assume honest users, and try to make Opacus default behavior privacy preserving (i.e. default to Poisson, secure_mode, etc.).
